VUFIND-1710 Add configurable separators for topics in alphabrowse

marktriggs opened the pull request:
Currently, subjects are shown differently between the record view in VuFind and the topic browse.  In the record view you see delimiters like this:

    Morris, William, > 1834-1896 > Criticism and interpretation.

But in the topic browse, the fields are just joined with spaces:

    Morris, William, 1834-1896 Criticism and interpretation.

This commit:

  * Adds a UTF-8 separator to delimit terms within topics in the Solr index

  * Adjusts the topic_browse definition to use a new TopicNormalizer, which strips out these new delimiters when building the sort key.

  * Changes the frontend of the AlphaBrowse to insert a user-configurable delimiter, shown when browsing topics.

Note: since Normalizer classes currently live in the vufind-browse-handler, there will be a corresponding commit there to add TopicNormalizer.

TODO
- [ ] Document Solr schema change when merging
- [x] Resolve [VUFIND-1710](https://openlibraryfoundation.atlassian.net/browse/VUFIND-1710) when merging

marktriggs committed:
VUFIND-1710 Add configurable separators for topics in alphabrowse

Currently, subjects are shown differently between the record view in
VuFind and the topic browse.  In the record view you see delimiters
like this:

    Morris, William, > 1834-1896 > Criticism and interpretation.

But in the topic browse, the fields are just joined with spaces:

    Morris, William, 1834-1896 Criticism and interpretation.

This commit:

  * Adds a UTF-8 separator to delimit terms within topics in the Solr
    index

  * Adjusts the topic_browse definition to use a new TopicNormalizer,
    which strips out these new delimiters when building the sort key.

  * Changes the frontend of the AlphaBrowse to insert a
    user-configurable delimiter, shown when browsing topics.

Note: since Normalizer classes currently live in the
vufind-browse-handler, there will be a corresponding commit there to
add TopicNormalizer.

marktriggs commented:
@demiankatz @dltj Just sending the pull request for VUFIND-1710 to get the review rolling :)

Note that there's also a vufind-browse-handler PR to go with this one: https://github.com/vufind-org/vufind-browse-handler/pull/51.  I wasn't sure how best to handle the bundled `browse-indexing.jar` and `browse-handler.jar` files, so I've left them out of this PR for now.

demiankatz commented:
> @demiankatz @dltj Just sending the pull request for VUFIND-1710 to get the review rolling :)
> 
> Note that there's also a vufind-browse-handler PR to go with this one: [vufind-org/vufind-browse-handler#51](https://github.com/vufind-org/vufind-browse-handler/pull/51). I wasn't sure how best to handle the bundled `browse-indexing.jar` and `browse-handler.jar` files, so I've left them out of this PR for now.

Thanks, @marktriggs! https://github.com/vufind-org/vufind-browse-handler/pull/51 looks reasonable to me, so I've gone ahead and merged it. That way, you can now build the .jar artifacts and add them to this PR so that we can test the whole thing together.

demiankatz commented on config/vufind/config.ini:
```diff
@@ -2156,13 +2156,21 @@ includeSchemaOrgMetadata = true
 
 ; The following two sections control the Alphabetic Browse module.
 [AlphaBrowse]
+
```
We don't usually have blanks at the start of sections; might be worth removing this one for consistency.

marktriggs replied:
Sorry, nervous habit :)

demiankatz commented on config/vufind/config.ini:
```diff
@@ -2156,13 +2156,21 @@ includeSchemaOrgMetadata = true
+
+; For topic browse, controls the separator to place between terms when
+; displaying topic headings.
+topic_browse_separator = " -- "
```
Since the value configured here differs from the default in the code, should this be commented out as an example? Should the existing default be mentioned in the comments?

marktriggs replied:
Gah, whoops.  This was me testing locally and it looked right enough when reviewing the diff that I didn't notice.  I've changed the line to show the default, commented it out and also clarified the leading comment to mention the default.

demiankatz commented on module/VuFind/src/VuFind/Controller/AlphabrowseController.php:
```diff
@@ -185,6 +190,24 @@ protected function addResultsToView(
+        $config = $this->getConfig();
+
+        foreach ($result['Browse']['items'] as &$item) {
+            $item['heading'] = str_replace("\u{2002}",
+                                           ($config->AlphaBrowse->topic_browse_separator ?? ' > '),
+                                           $item['heading']);
```
Suggested style change for consistency:

```suggestion
            $item['heading'] = str_replace(
                "\u{2002}",
                ($config->AlphaBrowse->topic_browse_separator ?? ' > '),
                $item['heading']
            );
```

marktriggs replied:
Done!

demiankatz reviewed (changes requested):
Here are a few comments from an initial review; I'll do more hands-on and automated testing once the .jar files have been added.

I'm also going to pin this to the 11.0 release, since 10.1 is very close and I don't want to introduce last-minute Solr schema changes.

marktriggs committed:
VUFIND-1710 Add the rebuilt vufind-browse-handler jars

Built from commit 25863e56d0fd4b02deedbfd72aca7a5c4885bdd7 in vufind-org/vufind-browse-handler

marktriggs committed:
VUFIND-1710 Apply suggestions from PR review

marktriggs commented:
Thanks @demiankatz,

I've applied the changes from your first review, and have checked in the updated `jar` files too.  Do you have a quick way of testing on Windows?  The `.bat` file changes are pretty trivial, but I'd feel better seeing them working.  I can try getting things running on a Windows machine here if you don't have one handy.

demiankatz commented on config/vufind/config.ini:
```diff
@@ -2158,11 +2158,19 @@ includeSchemaOrgMetadata = true
+
+; For topic browse, controls the separator to place between terms when
+; displaying topic headings.  Defaults to " > " if not set.
+;
+; topic_browse_separator = " > "
```
Small suggested edit to make this match our usual config style (spaces before human-readable comments, no spaces before example settings after comments):

```suggestion
;topic_browse_separator = " > "
```

demiankatz reviewed (changes requested):
Thanks, @marktriggs. I can try to test this out in Windows soon -- just having a busy week and haven't gotten that far yet.

In the meantime, see below for a tiny style suggestion.

Also, did you notice that tests are failing in PHP 8.3? Please let me know if you need any help investigating/sorting that out.

marktriggs committed:
VUFIND-1710 Fix mock data for unit test

The modified AlphaBrowse now makes use of sort_key, which is returned
by the browse handler.  This is basically the heading as stored in the
Solr index, as opposed to the 'heading' key which will have been
manipulated for display in some cases (such as Topic browse)

marktriggs committed:
VUFIND-1710 Fix comment style consistency

marktriggs committed:
VUFIND-1710 Fix doc comment syntax

marktriggs commented:
Thanks @demiankatz, and no rush from me.

I've fixed the comment syntax and figured out why that test was failing.  The AlphaBrowse code returns items with keys `heading` and `sort_key`, but the test code was mocking items and only including `heading`.

Historically that didn't matter because the browse frontend only used `heading` anyway.  There are two ways that the browse heading is used: the point where we show the heading to the user, and the point that we build a query link (with a `lookfor` parameter) to view the associated record(s).  Up until now, the heading shown to the user was identical to the value stored in the Solr index, so there was no distinction between the two cases (and that's why just using `heading` exclusively was fine).

But now we're rewriting the delimiter in topics, what we display to the user isn't the same as what's in the index.  So, `heading` is the label shown to the user, while `sort_key` is the index value needed for constructing a matching `lookfor` query.

demiankatz commented:
@marktriggs, I was able to confirm that this works correctly in Windows. I tested the dev branch: no separators; I switched to this branch and rebuilt the index: separators appeared!

demiankatz reviewed (approved):
This looks good to me, and all tests are passing in my local environment. I'll just leave this open a little longer in case @dltj or others have any last-minute suggestions or concerns.

dltj commented:
I'm good-to-go  with this.  Thanks, Mark and Demian!
